### DISCUSSION TITLE
how to use got lib or esm lib

### DISCUSSION THREAD

**Discussion by taotao7** [cawa-93/vite-electron-builder#856] [Q&A] [answered]
I'm having a problem with the goto request library, I don't know how to solve it, I tried to compile goto as esm myself but still can't run the project, thanks
![image](https://user-images.githubusercontent.com/24582204/196316725-321d839d-e895-405d-809e-c1f9db23f5f3.png)

**Comment by cawa-93** [comment, marked as answer]
Try add `got` to [ssr.noExternal](https://vitejs.dev/config/ssr-options.html#ssr-noexternal). It should include this package to `main` bundle and transpile it to cjs

**Reply by raphaelmenges** [reply]
I am facing the same issue: I am trying to use a ESM module in my project. I have tried to add the ESM module to `ssr.noExternal`:
```
ssr: { 
  noExternal: ["<module>"],
  format: "cjs",
},
```
However, the error persists for me:
```
App threw an error during load
Error [ERR_REQUIRE_ESM]: require() of ES Module [...]\app\node_modules\<module>\dist\index.js from [...]\app\packages\main\dist\index.cjs not supported.
```
I assume the idea to include the module in the transpiling will fix the issue. Perhaps I am doing the inclusion wrong?

**Reply by raphaelmenges** [reply]
I have created a minimal example extending the current template to debug the exact problem: <https://github.com/raphaelmenges/electron-vite4-esm>

As soon as I execute `npm run watch`, the bespoken `ERR_REQUIRE_ESM` error is thrown.

See changes to the template here: <https://github.com/raphaelmenges/electron-vite4-esm/commit/bcc5251079c9cc3ecd082d12599200ea1e7a35f9>

**Reply by cawa-93** [reply]
@raphaelmenges Seems you placed `ssr` config in wrong place. You placed top-level [`ssr`](https://vitejs.dev/config/ssr-options.html#ssr-options) options into [`build.ssr`](https://vitejs.dev/config/build-options.html#build-ssr). Correct should be:
```ts
const config = {

	// Settings for SSR
	ssr: {
	  noExternal: ['foo']
	},
	
	// Settings for build
	build: {
		ssr: true // Should the build context use the SSR mode
	}

}

**Reply by raphaelmenges** [reply]
This worked, thank you very much 👌
